File: prism-android-wallet/app/src/main/java/io/iohk/atala/prism/app/ui/main/credentials/CredentialDetailFragment.java
package io.iohk.atala.prism.app.ui.main.credentials;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Base64;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;

import javax.inject.Inject;

import io.iohk.atala.prism.app.ui.CvpFragment;
import io.iohk.cvp.R;
import io.iohk.atala.prism.app.data.local.db.model.Credential;
import io.iohk.atala.prism.app.utils.IntentDataConstants;
import io.iohk.atala.prism.app.ui.utils.AppBarConfigurator;
import io.iohk.atala.prism.app.ui.utils.StackedAppBar;
import io.iohk.cvp.databinding.FragmentCredentialDetailBinding;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Setter
@NoArgsConstructor
public class CredentialDetailFragment extends CvpFragment<CredentialDetailViewModel> {

    @Inject
    ViewModelProvider.Factory factory;

    private static final int REQUEST_DELETE_CREDENTIAL = 22;

    private FragmentCredentialDetailBinding binding;

    private String credentialId;

    /*
     * TODO this will be removed, replacing it with safeArgs
     * */
    public static CredentialDetailFragment build(String credentialId) {
        CredentialDetailFragment instance = new CredentialDetailFragment();
        instance.credentialId = credentialId;
        return instance;
    }

    @Override
    protected int getViewId() {
        return R.layout.fragment_credential_detail;
    }

    @Override
    protected AppBarConfigurator getAppBarConfigurator() {
        setHasOptionsMenu(true);
        return new StackedAppBar(R.string.home_title);
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        menu.findItem(R.id.action_credential_history).setVisible(true);
        menu.findItem(R.id.action_share_credential).setVisible(true);
        menu.findItem(R.id.action_delete_credential).setVisible(true);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {

        switch (item.getItemId()) {
            case R.id.action_share_credential:
                getShareFragment().show(requireActivity().getSupportFragmentManager(), null);
                return true;
            case android.R.id.home:
                requireActivity().onBackPressed();
                return true;
            case R.id.action_delete_credential:
                getDeleteCredentialFragment().show(requireActivity().getSupportFragmentManager(), null);
                return true;
            case R.id.action_credential_history:
                CredentialHistoryFragment fragment = CredentialHistoryFragment.Companion.build(credentialId);
                navigator.showFragmentOnTop(
                        requireActivity().getSupportFragmentManager(), fragment);
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private ShareCredentialDialogFragment getShareFragment() {
        ShareCredentialDialogFragment fragment = new ShareCredentialDialogFragment();
        Bundle args = new Bundle();
        args.putString(IntentDataConstants.CREDENTIAL_ID_KEY, credentialId);
        fragment.setArguments(args);
        return fragment;
    }

    private DeleteCredentialDialogFragment getDeleteCredentialFragment() {
        DeleteCredentialDialogFragment dialog = DeleteCredentialDialogFragment.Companion.build(credentialId);
        dialog.setTargetFragment(this, REQUEST_DELETE_CREDENTIAL);
        return dialog;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, getViewId(), container, false);
        binding.setLifecycleOwner(this);
        binding.setViewModel(getViewModel());
        setObservers();
        return binding.getRoot();
    }


    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getViewModel().fetchCredentialInfo(credentialId);
    }

    private void setObservers() {
        viewModel.getCredential().observe(getViewLifecycleOwner(), credential -> {
            fillWebView(credential);
            // TODO this is a hack to support the behavior of [CvpFragment] and the current custom navigation system, when we migrate to a native Android navigation and [CvpFragment] inheritance is removed this has to be deleted
            setActionBarTitle(CredentialUtil.getNameResource(credential.credentialType));
        });
    }

    private void fillWebView(Credential credential) {
        String credentialHtmlView = CredentialUtil.getHtml(credential);
        String encodedHtml = Base64.encodeToString(credentialHtmlView.getBytes(), Base64.NO_PADDING);
        binding.webView.loadData(encodedHtml, "text/html", "base64");
    }

    @Override
    public CredentialDetailViewModel getViewModel() {
        CredentialDetailViewModel viewModel = ViewModelProviders.of(this, factory)
                .get(CredentialDetailViewModel.class);
        return viewModel;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == REQUEST_DELETE_CREDENTIAL && resultCode == Activity.RESULT_OK) {
            requireActivity().onBackPressed();
        }
    }
}

File: prism-android-wallet/app/src/main/java/io/iohk/atala/prism/app/ui/main/credentials/CredentialDetailFragment.kt
package io.iohk.atala.prism.app.ui.main.credentials

import android.app.Activity
import android.os.Bundle
import android.util.Base64
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.setFragmentResultListener
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import dagger.android.support.DaggerFragment
import io.iohk.atala.prism.app.data.local.db.model.Credential
import io.iohk.atala.prism.app.neo.common.extensions.KEY_RESULT
import io.iohk.atala.prism.app.neo.common.extensions.supportActionBar
import io.iohk.cvp.R
import io.iohk.cvp.databinding.FragmentCredentialDetailBinding
import javax.inject.Inject

class CredentialDetailFragment : DaggerFragment() {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    private val viewModel: CredentialDetailViewModel by lazy {
        ViewModelProviders.of(this, viewModelFactory).get(CredentialDetailViewModel::class.java)
    }

    private val args: CredentialDetailFragmentArgs by navArgs()

    lateinit var binding: FragmentCredentialDetailBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        // Handle DeleteCredentialDialogFragment result
        setFragmentResultListener(DeleteCredentialDialogFragment.REQUEST_DELETE_CREDENTIAL) { _, bundle ->
            if (bundle.getInt(KEY_RESULT) == Activity.RESULT_OK) {
                findNavController().popBackStack()
            }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_credential_detail, container, false)
        binding.lifecycleOwner = this
        binding.viewModel = viewModel
        viewModel.fetchCredentialInfo(args.credentialId)
        setObservers()
        return binding.root
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.credential_detail_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_credential_history -> {
                val direction = CredentialDetailFragmentDirections.actionCredentialDetailFragmentToCredentialHistoryFragment(args.credentialId)
                findNavController().navigate(direction)
                true
            }
            R.id.action_delete_credential -> {
                val direction = CredentialDetailFragmentDirections.actionCredentialDetailFragmentToDeleteCredentialDialogFragment(args.credentialId)
                findNavController().navigate(direction)
                true
            }
            R.id.action_share_credential -> {
                val direction = CredentialDetailFragmentDirections.actionCredentialDetailFragmentToShareCredentialDialogFragment(args.credentialId)
                findNavController().navigate(direction)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun setObservers() {
        viewModel.credential.observe(viewLifecycleOwner) { credential ->
            fillWebView(credential)
            val title = getString(CredentialUtil.getNameResource(credential))
            findNavController().currentDestination?.label = title
            supportActionBar?.title = title
        }
    }

    private fun fillWebView(credential: Credential) {
        val credentialHtmlString = CredentialUtil.getHtml(credential)
        val encodedHtml = Base64.encodeToString(credentialHtmlString.toByteArray(), Base64.NO_PADDING)
        binding.webView.loadData(encodedHtml, "text/html", "base64")
    }
}
